dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusAArch64.java

```diff
@@ -105,163 +143,308 @@ public Object visitInstructionRow(LitmusAArch64Parser.InstructionRowContext ctx)
-        }
-        Load load = EventFactory.newRMWLoadExclusiveWithMo(register, address, ctx.loadExclusiveInstruction().mo);
-        return programBuilder.addChild(mainThread, load);
+    public Object visitLoadPair(LoadPairContext ctx) {
+        final boolean extended = ctx.rD064 != null;
+        final Register r064 = parseRegister64(ctx.rD032, ctx.rD064);
+        final Register r164 = parseRegister64(ctx.rD132, ctx.rD164);
+        final Register value0 = extended ? r064 : shrinkRegister(r064, ctx.rD032, false, false);
+        final Register value1 = extended ? r164 : shrinkRegister(r164, ctx.rD132, false, false);
+        final Expression address0 = parseAddress(ctx.address());
+        final Expression address1 = expressions.makeAdd(address0, expressions.makeValue(extended ? 8 : 4, i64));
+        add(EventFactory.newLoad(value0, address0));
+        add(EventFactory.newLoad(value1, address1));
+        addRegister64Update(r064, value0);
+        addRegister64Update(r164, value1);
+        return null;
     }
 
     @Override
-    public Object visitStore(LitmusAArch64Parser.StoreContext ctx) {
-        Register register = programBuilder.getOrNewRegister(mainThread, ctx.rV, archType);
-        Register address = programBuilder.getOrErrorRegister(mainThread, ctx.address().id);
-        if(ctx.offset() != null){
-            address = visitOffset(ctx.offset(), address);
-        }
-        return programBuilder.addChild(mainThread, EventFactory.newStoreWithMo(address, register, ctx.storeInstruction().mo));
+    public Object visitLoadExclusive(LoadExclusiveContext ctx) {
+        final Register r64 = parseRegister64(ctx.rD32, ctx.rD64);
+        final LoadExclusiveInstructionContext inst = ctx.loadExclusiveInstruction();
+        final Register register = shrinkRegister(r64, ctx.rD32, inst.halfWordSize, inst.byteSize);
+        final Expression address = parseAddress(ctx.address());
+        final String mo = inst.acquire ? MO_ACQ : "";
+        add(EventFactory.newRMWLoadExclusiveWithMo(register, address, mo));
+        addRegister64Update(r64, register);
+        return null;
     }
 
     @Override
-    public Object visitStoreExclusive(LitmusAArch64Parser.StoreExclusiveContext ctx) {
-        Register register = programBuilder.getOrNewRegister(mainThread, ctx.rV, archType);
-        Register statusReg = programBuilder.getOrNewRegister(mainThread, ctx.rS, archType);
-        Register address = programBuilder.getOrErrorRegister(mainThread, ctx.address().id);
-        if(ctx.offset() != null){
-            address = visitOffset(ctx.offset(), address);
+    public Object visitStore(StoreContext ctx) {
+        final Register r64 = parseRegister64(ctx.rV32, ctx.rV64);
+        final StoreInstructionContext inst = ctx.storeInstruction();
+        final IntegerType type = inst.byteSize ? i8 : inst.halfWordSize ? i16 : i32;
+        final Expression value = ctx.rV64 != null ? r64 : expressions.makeIntegerCast(r64, type, false);
```

Comment:
Why not always use the cast and let the (yet to come) `ExpSimplifier` get rid of the cast for the 64-bits case. Same comment applies to other parts